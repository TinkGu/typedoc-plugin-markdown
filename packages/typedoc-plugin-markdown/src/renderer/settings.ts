import { DeclarationReflection, ProjectReflection } from 'typedoc';

import { PluginOptions } from './types';

let _project: ProjectReflection;
let _options: PluginOptions;
let _activeUrl: string;
let _activeReflection: DeclarationReflection;

export default {
  get project() {
    return _project;
  },

  set project(project: ProjectReflection) {
    _project = project;
  },

  get options() {
    return _options;
  },

  set options(options: PluginOptions) {
    _options = options;
  },

  get activeUrl() {
    return _activeUrl;
  },

  set activeUrl(url) {
    _activeUrl = url;
  },

  get activeReflection() {
    return _activeReflection;
  },

  set activeReflection(reflection) {
    _activeReflection = reflection;
  },

  get globalsFile() {
    return 'modules.md';
  },

  get globalsName() {
    return _options.entryPoints && _options.entryPoints.length > 1
      ? 'Modules'
      : 'Exports';
  },
};
