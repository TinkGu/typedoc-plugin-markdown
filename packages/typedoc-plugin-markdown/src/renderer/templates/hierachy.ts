import { DeclarationHierarchy } from 'typedoc';
import { spaces } from '../tools/utils';
import { typeTemplate } from './type';

export function hierarchyTemplate(model: DeclarationHierarchy) {
  return hierarchy(model);
}

function hierarchy(model: DeclarationHierarchy, level = 0) {
  const md: string[] = [];
  const symbol = level > 0 ? getSymbol(level) : '-';
  model.types.forEach((hierarchyType) => {
    if (model.isTarget) {
      md.push(`${symbol} **\`${hierarchyType}\`**`);
    } else {
      md.push(`${symbol} ${typeTemplate(hierarchyType)}`);
    }
  });
  if (model.next) {
    md.push(hierarchy(model.next, level + 1));
  }
  return md.join('\n\n');
}

function getSymbol(level: number) {
  return spaces(2) + [...Array(level)].map(() => 'â†³').join('');
}
