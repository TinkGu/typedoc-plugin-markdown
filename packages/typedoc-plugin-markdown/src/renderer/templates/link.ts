import * as path from 'path';

import settings from '../settings';
import { link } from '../tools/elements';
import { escapeChars } from '../tools/utils';

export function linkTemplate(label: string, to = '', escapeLabel = true) {
  label = escapeLabel ? escapeChars(label) : label;
  return link(label, getUrl(to));
}

function getUrl(to: string) {
  const { options, activeUrl } = settings;

  const urlPrefix = /^(http|ftp)s?:\/\//;
  if (urlPrefix.test(to)) {
    return to;
  } else {
    return options.publicPath
      ? getPublicUrl(options.publicPath, to)
      : getRelativeUrl(activeUrl, to);
  }
}

function getPublicUrl(publicPath: string, url: string) {
  return publicPath + url;
}

function getRelativeUrl(from: string, to: string) {
  const relative = path.relative(path.dirname(from), path.dirname(to));
  return path.join(relative, path.basename(to)).replace(/\\/g, '/');
}
