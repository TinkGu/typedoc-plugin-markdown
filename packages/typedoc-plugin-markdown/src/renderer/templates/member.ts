import { DeclarationReflection, ReferenceReflection } from 'typedoc';
import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { heading } from '../tools/elements';
import { escapeChars } from '../tools/utils';
import { declarationTemplate } from './declaration';
import { signatureTemplate } from './signature';

export function memberTemplate(
  model: DeclarationReflection | ReferenceReflection,
) {
  const options = settings.options;
  const md = new MarkdownBuilder();

  md.add(
    heading(
      3,
      options.namedAnchors
        ? `<a id="${model.anchor}" name="${model.anchor}"></a> `
        : '' + escapeChars(model.name),
    ),
  );

  if (model.signatures) {
    model.signatures.forEach((signature) => {
      md.add(signatureTemplate(signature));
    });
  } else {
    if (model.hasGetterOrSetter()) {
      if (model.getSignature) {
        md.add(signatureTemplate(model.getSignature, false, 'get'));
      }
      if (model.setSignature) {
        md.add(signatureTemplate(model.setSignature, false, 'set'));
      }
    } else {
      if (model instanceof ReferenceReflection && model.isReference) {
        const deep = model.tryGetTargetReflectionDeep();
        if (deep) {
          md.add(`Re-exports ${deep.name}`);
        }
      }

      if (model instanceof DeclarationReflection) {
        md.add(declarationTemplate(model));
      }
    }
  }

  return md.print();
}
