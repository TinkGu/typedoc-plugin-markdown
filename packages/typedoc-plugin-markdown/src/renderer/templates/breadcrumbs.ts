import { Reflection } from 'typedoc';
import { PageEvent } from 'typedoc/dist/lib/output/events';
import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { escapeChars } from '../tools/utils';
import { linkTemplate } from './link';

export function breadcrumbsTemplate(page: PageEvent) {
  const { activeUrl, globalsName, globalsFile, options } = settings;

  const md = new MarkdownBuilder();

  md.add(
    page.url === options.entryDocument
      ? page.project.name
      : linkTemplate(page.project.name, options.entryDocument),
  );

  if (!options.readme.endsWith('none')) {
    if (activeUrl === globalsFile) {
      md.add(globalsName);
    } else {
      md.add(linkTemplate(globalsName, globalsFile));
    }
  }

  const addBreadcrumb = (model: Reflection, md: MarkdownBuilder) => {
    if (model && model.parent) {
      addBreadcrumb(model.parent, md);

      if (activeUrl === model.url) {
        md.add(escapeChars(model.name));
      } else {
        md.add(linkTemplate(model.name, model.url));
      }
    }
  };

  addBreadcrumb(page.model, md);

  return md.print(' / ');
}
