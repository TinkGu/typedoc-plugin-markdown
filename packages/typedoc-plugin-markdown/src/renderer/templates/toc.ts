import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';

import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { heading, unorderedList } from '../tools/elements';
import { linkTemplate } from './link';

export function tocTemplate(model: ProjectReflection | DeclarationReflection) {
  const options = settings.options;
  const md = new MarkdownBuilder();
  const isVisible = model.groups?.some((group) =>
    group.allChildrenHaveOwnDocument(),
  );
  if ((!this.hideInPageTOC && model.groups) || (isVisible && model.groups)) {
    if (!options.hideInPageTOC) {
      md.add(heading(2, 'Table of contents'));
    }
    model.groups
      ?.filter(
        (group) => !options.hideInPageTOC || group.allChildrenHaveOwnDocument(),
      )
      .forEach((group) => {
        if (group.categories) {
          group.categories.forEach((category) => {
            md.add(getGroup(category, category.title + ' ' + group.title));
          });
        } else {
          md.add(getGroup(group, group.title));
        }
      });
  }

  return md.print();
}

function getGroup(group: ReflectionGroup | ReflectionCategory, title: string) {
  const options = settings.options;
  const md = new MarkdownBuilder();
  md.add(heading(options.hideInPageTOC ? 2 : 3, title));
  const groupItems = group.children.map((child) =>
    linkTemplate(child.name, child.url),
  );
  md.add(unorderedList(groupItems));
  return md.print();
}
