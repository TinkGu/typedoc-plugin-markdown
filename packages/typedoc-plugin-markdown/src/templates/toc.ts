import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';
import { getContext } from '../context';
import { MarkdownBuilder } from '../tools/builder';
import { heading, unorderedList } from '../tools/elements';
import { linkTemplate } from './link';

export function tocTemplate(model: ProjectReflection | DeclarationReflection) {
  const { hideInPageTOC } = getContext();

  const md = new MarkdownBuilder();
  const isVisible = model.groups?.some((group) =>
    group.allChildrenHaveOwnDocument(),
  );
  if ((!this.hideInPageTOC && model.groups) || (isVisible && model.groups)) {
    if (!hideInPageTOC) {
      md.add(heading(2, 'Table of contents'));
    }
    model.groups
      ?.filter((group) => !hideInPageTOC || group.allChildrenHaveOwnDocument())
      .forEach((group) => {
        if (group.categories) {
          group.categories.forEach((category) => {
            md.add(getGroup(category, category.title + ' ' + group.title));
          });
        } else {
          md.add(getGroup(group, group.title));
        }
      });
  }

  return md.print();
}

function getGroup(group: ReflectionGroup | ReflectionCategory, title: string) {
  const { hideInPageTOC } = getContext();
  const md = new MarkdownBuilder();
  md.add(heading(hideInPageTOC ? 2 : 3, title));
  const groupItems = group.children.map((child) =>
    linkTemplate(child.name, child.url),
  );
  md.add(unorderedList(groupItems));
  return md.print();
}
