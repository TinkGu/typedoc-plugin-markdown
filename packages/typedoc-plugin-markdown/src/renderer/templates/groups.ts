import {
  DeclarationReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';

import { MarkdownBuilder } from '../tools/builder';
import { heading, horizontalRule } from '../tools/elements';
import { memberTemplate } from './member';

export function groupsTemplate(groups: ReflectionGroup[]) {
  const md = new MarkdownBuilder();

  groups
    ?.filter((group) => !group.allChildrenHaveOwnDocument())
    .forEach((group) => {
      if (group.categories) {
        group.categories
          ?.filter((category) => !category.allChildrenHaveOwnDocument())
          .forEach((category) => {
            md.add(heading(2, `${group.title} ${category.title}`));
            md.add(getMembers(category));
          });
      } else {
        md.add(heading(2, group.title));
        md.add(getMembers(group));
      }
    });

  return md.print();
}

function getMembers(item: ReflectionGroup | ReflectionCategory) {
  const md = new MarkdownBuilder();
  item.children.forEach((child, index) => {
    md.add(memberTemplate(child as DeclarationReflection));
    if (index !== item.children.length - 1) {
      md.add(horizontalRule());
    }
  });
  return md.print();
}
