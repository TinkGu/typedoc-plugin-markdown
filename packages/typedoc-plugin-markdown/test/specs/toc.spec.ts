import { Reflection } from 'typedoc';
import settings from '../../src/renderer/settings';
import { tocTemplate } from '../../src/renderer/templates/toc';
import { TestApp } from '../test-app';

describe(`TOC:`, () => {
  let moduleReflection: Reflection;
  let classReflection: Reflection;

  describe(`(default)`, () => {
    let testApp: TestApp;
    beforeAll(() => {
      testApp = new TestApp(['breadcrumbs.ts']);
      testApp.bootstrap();
      moduleReflection = testApp.project.children[0];
      classReflection = testApp.project.findReflectionByName('Breadcrumbs');
      settings.activeUrl = 'classes/Breadcrumbs.md';
    });

    afterAll(() => {
      testApp.cleanup();
    });

    test(`should display toc for module'`, () => {
      expect(tocTemplate(moduleReflection as any)).toMatchSnapshot();
    });

    test(`should display toc for class'`, () => {
      expect(tocTemplate(classReflection as any)).toMatchSnapshot();
    });
  });
  describe(`(hideInPageToc)`, () => {
    let testApp: TestApp;
    beforeAll(() => {
      testApp = new TestApp(['breadcrumbs.ts']);
      testApp.bootstrap({ hideInPageTOC: true });
      moduleReflection = testApp.project.children[0];
      classReflection = testApp.project.findReflectionByName('Breadcrumbs');
      settings.activeUrl = 'classes/Breadcrumbs.md';
    });

    afterAll(() => {
      testApp.cleanup();
    });

    test(`should not display toc for class'`, () => {
      expect(tocTemplate(classReflection as any)).toEqual('');
    });
  });
});
