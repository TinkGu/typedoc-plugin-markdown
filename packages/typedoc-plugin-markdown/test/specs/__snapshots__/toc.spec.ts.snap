// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TOC: (default) should display toc for class' 1`] = `
"## Table of contents

### Constructors

- [constructor](Breadcrumbs.md#constructor)

### Methods

- [someMethod](Breadcrumbs.md#somemethod)"
`;

exports[`TOC: (default) should display toc for module' 1`] = `
"## Table of contents

### Constructors

- [constructor](Breadcrumbs.md#constructor)

### Methods

- [someMethod](Breadcrumbs.md#somemethod)"
`;
